File: packages/go/storage/storage_test.go
package storage

import (
	"context"
	"testing"
)

type fakeStorage struct{}

func (fakeStorage) Reserve(context.Context, ReferenceReservation) (bool, error) {
	return true, nil
}

func (fakeStorage) Resolve(_ context.Context, _ string, namespace string) (*ReferenceMapping, error) {
	return &ReferenceMapping{MachineID: "01890f8c-7b2a-7cc3-98b0-112233445566", Namespace: namespace}, nil
}

func (fakeStorage) Allocate(context.Context, SequenceAllocationRequest) (uint64, error) {
	return 1, nil
}

func TestAdapterExposesAllStorageOperations(t *testing.T) {
	var adapter Adapter = fakeStorage{}
	reservation := ReferenceReservation{
		MachineID: "01890f8c-7b2a-7cc3-98b0-112233445566",
		Namespace: "order",
		Reference: "ORD-0123-4567-89-P",
	}
	reserved, err := adapter.Reserve(context.Background(), reservation)
	if err != nil || !reserved {
		t.Fatalf("reserve failed: reserved=%v err=%v", reserved, err)
	}
	mapping, err := adapter.Resolve(context.Background(), reservation.Reference, reservation.Namespace)
	if err != nil || mapping == nil || mapping.MachineID != reservation.MachineID {
		t.Fatalf("resolve failed: mapping=%#v err=%v", mapping, err)
	}
	sequence, err := adapter.Allocate(context.Background(), SequenceAllocationRequest{
		MachineID: reservation.MachineID, Namespace: "receipt", ReferencePrefix: "RCT", Width: 4,
	})
	if err != nil || sequence != 1 {
		t.Fatalf("allocate failed: sequence=%d err=%v", sequence, err)
	}
}

File: packages/typescript/test/storage.test.ts
import { expect, test } from "vitest";

import type { HumanReference, MachineId } from "../src/index.js";
import { createStorageAdapter } from "../src/storage.js";

const machineId = "01890f8c-7b2a-7cc3-98b0-112233445566" as MachineId;
const reference = "ORD-0123-4567-89-P" as HumanReference;

test("a storage adapter exposes reservation, resolution, and allocation", async () => {
  const adapter = createStorageAdapter({
    referenceStore: { reserve: () => Promise.resolve(true) },
    lookup: () => Promise.resolve({ machineId, namespace: "order" }),
    sequenceAllocator: { allocate: () => Promise.resolve(1n) },
  });

  await expect(
    adapter.referenceStore.reserve({
      machineId,
      namespace: "order",
      reference,
    }),
  ).resolves.toBe(true);
  await expect(adapter.lookup(reference, "order")).resolves.toEqual({
    machineId,
    namespace: "order",
  });
  await expect(
    adapter.sequenceAllocator.allocate({
      machineId,
      namespace: "receipt",
      referencePrefix: "RCT",
      scope: null,
      width: 4,
    }),
  ).resolves.toBe(1n);
});
